File: dashboard_data_generator.py
import json
from log_parser import NginxLogParser
import os
import argparse
from datetime import datetime
from collections import Counter, defaultdict

def generate_dashboard_data(log_file, output_dir):
    """Generate formatted data for the Next.js dashboard."""
    # Parse the log file
    parser = NginxLogParser(log_file)
    parser.parse_logs()
    stats = parser.get_statistics()
    
    # Create the output directory if it doesn't exist
    os.makedirs(output_dir, exist_ok=True)
    
    # Generate data files for different visualizations
    
    # 1. Status code distribution
    status_data = [
        {"status": str(status), "count": count}
        for status, count in stats['status_distribution'].items()
    ]
    with open(os.path.join(output_dir, 'status_codes.json'), 'w') as f:
        json.dump(status_data, f, indent=2)
    
    # 2. Requests timeline (hourly)
    timeline_data = [
        {"hour": hour, "count": count}
        for hour, count in stats['requests_timeline']
    ]
    with open(os.path.join(output_dir, 'requests_timeline.json'), 'w') as f:
        json.dump(timeline_data, f, indent=2)
    
    # 3. Daily requests timeline
    daily_data = [
        {"date": day, "count": count}
        for day, count in stats['daily_requests']
    ]
    with open(os.path.join(output_dir, 'daily_requests.json'), 'w') as f:
        json.dump(daily_data, f, indent=2)
    
    # 4. Traffic per hour (in MB)
    traffic_data = [
        {"hour": hour, "megabytes": round(mb, 2)}
        for hour, mb in stats['traffic_timeline_mb']
    ]
    with open(os.path.join(output_dir, 'traffic_timeline.json'), 'w') as f:
        json.dump(traffic_data, f, indent=2)
    
    # 5. Top endpoints
    endpoint_data = [
        {"endpoint": endpoint, "count": count}
        for endpoint, count in stats['top_endpoints']  # All top endpoints
    ]
    with open(os.path.join(output_dir, 'top_endpoints.json'), 'w') as f:
        json.dump(endpoint_data, f, indent=2)
    
    # 6. Top IP addresses
    ip_data = [
        {"ip": ip, "count": count}
        for ip, count in stats['top_ips']  # All top IPs
    ]
    with open(os.path.join(output_dir, 'top_ips.json'), 'w') as f:
        json.dump(ip_data, f, indent=2)
    
    # 7. Top user agents
    user_agent_data = [
        {"user_agent": user_agent, "count": count}
        for user_agent, count in stats['top_user_agents']
    ]
    with open(os.path.join(output_dir, 'top_user_agents.json'), 'w') as f:
        json.dump(user_agent_data, f, indent=2)
    
    # 8. File type distribution
    file_type_data = [
        {"type": file_type, "count": count}
        for file_type, count in stats['top_file_types']
    ]
    with open(os.path.join(output_dir, 'file_types.json'), 'w') as f:
        json.dump(file_type_data, f, indent=2)
    
    # 9. Bot vs User distribution
    bot_user_data = [
        {"type": "Bot", "count": stats['bot_requests'], "percentage": round(stats['bot_percentage'], 1)},
        {"type": "User", "count": stats['user_requests'], "percentage": round(stats['user_percentage'], 1)}
    ]
    with open(os.path.join(output_dir, 'bot_user.json'), 'w') as f:
        json.dump(bot_user_data, f, indent=2)
    
    # 10. HTTP methods distribution
    method_data = [
        {"method": method, "count": count}
        for method, count in stats['top_http_methods']
    ]
    with open(os.path.join(output_dir, 'http_methods.json'), 'w') as f:
        json.dump(method_data, f, indent=2)
    
    # 11. Path depth distribution
    path_depth_data = [
        {"depth": depth, "count": count}
        for depth, count in stats['path_depth_distribution']
    ]
    with open(os.path.join(output_dir, 'path_depth.json'), 'w') as f:
        json.dump(path_depth_data, f, indent=2)
    
    # 12. Protocol version distribution
    protocol_data = [
        {"protocol": protocol, "count": count}
        for protocol, count in stats['protocol_distribution'].items()
    ]
    with open(os.path.join(output_dir, 'protocols.json'), 'w') as f:
        json.dump(protocol_data, f, indent=2)
    
    # 13. Top error paths with enhanced data
    # Extract common status codes for each error path
    error_paths_with_status = []
    for path, count in stats['top_error_paths']:
        # Find the most common status code for this path
        path_status_counts = Counter()
        for log in parser.parsed_logs:
            if log.get('path') == path and int(log.get('status', 0)) >= 400:
                path_status_counts[log.get('status')] += 1
        
        # Get the most common status code
        most_common_status = path_status_counts.most_common(1)[0][0] if path_status_counts else 'Unknown'
        
        error_paths_with_status.append({
            "path": path, 
            "count": count,
            "status": most_common_status,
            "status_distribution": dict(path_status_counts)
        })
    
    with open(os.path.join(output_dir, 'error_paths.json'), 'w') as f:
        json.dump(error_paths_with_status, f, indent=2)
    
    # 14. IP classification
    ip_class_data = [
        {"class": ip_class, "count": count}
        for ip_class, count in stats['ip_classification'].items()
    ]
    with open(os.path.join(output_dir, 'ip_classification.json'), 'w') as f:
        json.dump(ip_class_data, f, indent=2)
    
    # 15. Top referrers
    referrer_data = [
        {"referrer": referrer, "count": count}
        for referrer, count in stats['top_referers']
    ]
    with open(os.path.join(output_dir, 'top_referrers.json'), 'w') as f:
        json.dump(referrer_data, f, indent=2)
    
    # 16. Status code category distribution
    status_category_data = [
        {"category": "Success (2xx)", "count": stats['success_requests'], "percentage": round(stats['success_percentage'], 1)},
        {"category": "Redirect (3xx)", "count": stats['redirect_requests'], "percentage": round(stats['redirect_percentage'], 1)},
        {"category": "Client Error (4xx)", "count": stats['client_error_requests'], "percentage": round(stats['client_error_percentage'], 1)},
        {"category": "Server Error (5xx)", "count": stats['server_error_requests'], "percentage": round(stats['server_error_percentage'], 1)}
    ]
    with open(os.path.join(output_dir, 'status_categories.json'), 'w') as f:
        json.dump(status_category_data, f, indent=2)
    
    # 17. Complete dataset for filtering (limit to first 5000 entries for performance)
    filtered_logs = []
    for log in parser.parsed_logs[:5000]:  # Limit for performance
        filtered_log = {
            "timestamp": log.get('datetime_obj').isoformat() if log.get('datetime_obj') else None,
            "ip": log.get('ip'),
            "method": log.get('method'),
            "path": log.get('path'),
            "status": log.get('status'),
            "bytes": log.get('bytes'),
            "user_agent": log.get('user_agent'),
            "is_bot": log.get('is_bot'),
            "referrer": log.get('referrer'),
            "file_extension": log.get('file_extension'),
            "hour": log.get('hour')
        }
        filtered_logs.append(filtered_log)
    
    with open(os.path.join(output_dir, 'filtered_logs.json'), 'w') as f:
        json.dump(filtered_logs, f, indent=2)
    
    # 18. Summary statistics with enhanced metrics
    summary_data = {
        "total_requests": stats['total_requests'],
        "total_transferred_mb": round(stats['total_transferred_mb'], 2),
        "unique_ips": len(stats['ip_addresses']),
        "unique_endpoints": len(stats['endpoints']),
        "bot_percentage": round(stats['bot_percentage'], 1),
        "success_percentage": round(stats['success_percentage'], 1),
        "error_percentage": round(stats['client_error_percentage'] + stats['server_error_percentage'], 1),
        "most_common_status": str(stats['status_codes'].most_common(1)[0][0]) if stats['status_codes'] else "None",
        "most_common_method": stats['top_http_methods'][0][0] if stats['top_http_methods'] else "None",
        "most_popular_endpoint": stats['top_endpoints'][0][0] if stats['top_endpoints'] else "None",
        "generated_at": datetime.now().isoformat()
    }
    with open(os.path.join(output_dir, 'summary.json'), 'w') as f:
        json.dump(summary_data, f, indent=2)
        
    # 19. Raw statistics for advanced filtering
    with open(os.path.join(output_dir, 'raw_stats.json'), 'w') as f:
        # Need to convert some data structures for JSON serialization
        json_safe_stats = {}
        for key, value in stats.items():
            if isinstance(value, (Counter, defaultdict)):
                json_safe_stats[key] = dict(value)
            else:
                json_safe_stats[key] = value
        json.dump(json_safe_stats, f, indent=2)
    
    print(f"Generated dashboard data files in {output_dir}")
    return output_dir

def main():
    parser = argparse.ArgumentParser(description='Generate dashboard data from NGINX logs')
    parser.add_argument('log_file', help='Path to the NGINX access log file (can be gzipped)')
    parser.add_argument('--output-dir', '-o', help='Output directory for dashboard data files', 
                        default='nginx-dashboard/public/data')
    
    args = parser.parse_args()
    
    if not os.path.exists(args.log_file):
        print(f"Error: Log file '{args.log_file}' not found")
        return
    
    generate_dashboard_data(args.log_file, args.output_dir)

if __name__ == "__main__":
    main()

File: dashboard_data_generator.js
/**
 * NGINX Log Dashboard - Data Generator
 * 
 * This script processes NGINX logs and generates JSON files for dashboard visualization.
 * It replaces the Python-based solution with a JavaScript implementation.
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');
const { createGzip } = require('zlib');
const { promisify } = require('util');

// Configuration
const config = {
  logFile: path.resolve(__dirname, 'access.log'),
  outputDir: path.resolve(__dirname, 'nginx-dashboard/public/data'),
  sampleSize: 10000, // Number of log entries to process (0 for all entries)
};

// Parse NGINX log line (combined format)
function parseNginxLog(line) {
  // Match NGINX combined log format
  const regex = /^(\S+) - - \[([^\]]+)\] "(\S+) (\S+) (\S+)" (\d+) (\d+) "([^"]*)" "([^"]*)"/;
  const match = line.match(regex);
  
  if (!match) {
    return null;
  }
  
  const [_, ip, timeStr, method, path, protocol, status, bytes, referrer, userAgent] = match;
  
  // Convert timestamp (format: "10/Oct/2023:13:55:36 +0000")
  const timestamp = new Date(timeStr.replace(':', ' '));
  
  // Simple bot detection based on user agent
  const isBot = 
    /bot|crawl|spider|preview|scan|check|index|monitor/i.test(userAgent) || 
    /googlebot|bingbot|yandexbot|baiduspider|facebookexternalhit/i.test(userAgent);
  
  return {
    ip,
    timestamp,
    method,
    path,
    protocol,
    status: parseInt(status),
    bytes: parseInt(bytes),
    referrer: referrer === '-' ? null : referrer,
    user_agent: userAgent,
    is_bot: isBot
  };
}

// Calculate statistics from parsed logs
function calculateStatistics(logs) {
  console.log(`Calculating statistics for ${logs.length} log entries...`);
  
  // Basic summary statistics
  const totalRequests = logs.length;
  const uniqueIPs = new Set(logs.map(log => log.ip)).size;
  const errorRequests = logs.filter(log => log.status >= 400).length;
  const botRequests = logs.filter(log => log.is_bot).length;
  const totalBytes = logs.reduce((sum, log) => sum + log.bytes, 0);
  
  // Status code statistics
  const statusCodes = {};
  logs.forEach(log => {
    const status = log.status.toString();
    statusCodes[status] = (statusCodes[status] || 0) + 1;
  });
  
  // HTTP methods statistics
  const httpMethods = {};
  logs.forEach(log => {
    const method = log.method;
    httpMethods[method] = (httpMethods[method] || 0) + 1;
  });
  
  // Timeline statistics (hourly)
  const hourlyTimeline = {};
  logs.forEach(log => {
    const date = log.timestamp.toISOString().split('T')[0];
    const hour = log.timestamp.getHours();
    const key = `${date} ${hour}:00`;
    hourlyTimeline[key] = (hourlyTimeline[key] || 0) + 1;
  });
  
  // Timeline statistics (daily)
  const dailyTimeline = {};
  logs.forEach(log => {
    const date = log.timestamp.toISOString().split('T')[0];
    dailyTimeline[date] = (dailyTimeline[date] || 0) + 1;
  });
  
  // Traffic data (bytes per hour)
  const hourlyTraffic = {};
  logs.forEach(log => {
    const date = log.timestamp.toISOString().split('T')[0];
    const hour = log.timestamp.getHours();
    const key = `${date} ${hour}:00`;
    hourlyTraffic[key] = (hourlyTraffic[key] || 0) + log.bytes;
  });
  
  // Top IPs
  const ipCounts = {};
  logs.forEach(log => {
    ipCounts[log.ip] = (ipCounts[log.ip] || 0) + 1;
  });
  
  // Top endpoints
  const endpointCounts = {};
  logs.forEach(log => {
    endpointCounts[log.path] = (endpointCounts[log.path] || 0) + 1;
  });
  
  // Top referrers
  const referrerCounts = {};
  logs.forEach(log => {
    if (log.referrer) {
      referrerCounts[log.referrer] = (referrerCounts[log.referrer] || 0) + 1;
    }
  });
  
  // Bot vs user data
  const botUserData = [
    { type: 'Bot', count: botRequests },
    { type: 'User', count: totalRequests - botRequests }
  ];
  
  // Status categories
  const statusCategories = [
    { category: '2xx', count: logs.filter(log => log.status >= 200 && log.status < 300).length },
    { category: '3xx', count: logs.filter(log => log.status >= 300 && log.status < 400).length },
    { category: '4xx', count: logs.filter(log => log.status >= 400 && log.status < 500).length },
    { category: '5xx', count: logs.filter(log => log.status >= 500 && log.status < 600).length }
  ];
  
  // File types
  const fileTypes = {};
  logs.forEach(log => {
    const match = log.path.match(/\.([^.?]+)(?:\?.*)?$/);
    const fileType = match ? match[1] : 'none';
    fileTypes[fileType] = (fileTypes[fileType] || 0) + 1;
  });
  
  // Path depth
  const pathDepth = {};
  logs.forEach(log => {
    const depth = log.path.split('/').filter(Boolean).length;
    pathDepth[depth] = (pathDepth[depth] || 0) + 1;
  });
  
  // Error paths
  const errorPaths = {};
  logs.filter(log => log.status >= 400).forEach(log => {
    const key = `${log.path} (${log.status})`;
    errorPaths[key] = (errorPaths[key] || 0) + 1;
  });
  
  // Format the data for the dashboard
  return {
    summary: [
      { metric: 'Total Requests', value: totalRequests },
      { metric: 'Unique IPs', value: uniqueIPs },
      { metric: 'Error Requests', value: errorRequests },
      { metric: 'Bot Requests', value: botRequests },
      { metric: 'Total Bytes', value: totalBytes }
    ],
    statusCodes: Object.entries(statusCodes).map(([status, count]) => ({ status, count })).sort((a, b) => b.count - a.count),
    httpMethods: Object.entries(httpMethods).map(([method, count]) => ({ method, count })).sort((a, b) => b.count - a.count),
    requestsTimeline: Object.entries(hourlyTimeline).map(([time, count]) => {
      const [date, hour] = time.split(' ');
      return { hour, count, fullTime: time };
    }).sort((a, b) => a.fullTime.localeCompare(b.fullTime)),
    dailyRequests: Object.entries(dailyTimeline).map(([date, count]) => ({
      date,
      count,
      fullDate: date
    })).sort((a, b) => a.fullDate.localeCompare(b.fullDate)),
    trafficTimeline: Object.entries(hourlyTraffic).map(([time, bytes]) => {
      const [date, hour] = time.split(' ');
      return { hour, megabytes: bytes / (1024 * 1024), fullTime: time };
    }).sort((a, b) => a.fullTime.localeCompare(b.fullTime)),
    topIps: Object.entries(ipCounts).map(([ip, count]) => ({
      ip,
      count,
      botPercentage: (logs.filter(log => log.ip === ip && log.is_bot).length / ipCounts[ip]) * 100
    })).sort((a, b) => b.count - a.count).slice(0, 100),
    topEndpoints: Object.entries(endpointCounts).map(([endpoint, count]) => ({ endpoint, count })).sort((a, b) => b.count - a.count).slice(0, 100),
    topReferrers: Object.entries(referrerCounts).map(([referrer, count]) => ({ referrer, count })).sort((a, b) => b.count - a.count).slice(0, 100),
    botUser: botUserData,
    statusCategories: statusCategories,
    fileTypes: Object.entries(fileTypes).map(([type, count]) => ({ type, count })).sort((a, b) => b.count - a.count).slice(0, 20),
    pathDepth: Object.entries(pathDepth).map(([depth, count]) => ({ depth: parseInt(depth), count })).sort((a, b) => a.depth - b.depth),
    errorPaths: Object.entries(errorPaths).map(([path, count]) => {
      const match = path.match(/^(.*) \((\d+)\)$/);
      return { path: match[1], count, status: match[2] };
    }).sort((a, b) => b.count - a.count).slice(0, 100),
    filteredLogs: logs.map(log => ({
      timestamp: log.timestamp.toISOString(),
      ip: log.ip,
      method: log.method,
      path: log.path,
      protocol: log.protocol,
      status: log.status.toString(),
      bytes: log.bytes,
      referrer: log.referrer,
      user_agent: log.user_agent,
      is_bot: log.is_bot
    })).slice(0, 1000)
  };
}

// Save statistics to JSON files
async function saveStatistics(stats) {
  console.log('Saving statistics to JSON files...');
  
  // Ensure output directory exists
  if (!fs.existsSync(config.outputDir)) {
    fs.mkdirSync(config.outputDir, { recursive: true });
  }
  
  const files = {
    'summary.json': stats.summary,
    'status_codes.json': stats.statusCodes,
    'http_methods.json': stats.httpMethods,
    'requests_timeline.json': stats.requestsTimeline,
    'daily_requests.json': stats.dailyRequests,
    'traffic_timeline.json': stats.trafficTimeline,
    'top_ips.json': stats.topIps,
    'top_endpoints.json': stats.topEndpoints,
    'top_referrers.json': stats.topReferrers,
    'bot_user.json': stats.botUser,
    'status_categories.json': stats.statusCategories,
    'file_types.json': stats.fileTypes,
    'path_depth.json': stats.pathDepth,
    'error_paths.json': stats.errorPaths,
    'filtered_logs.json': stats.filteredLogs,
    'raw_stats.json': stats // Complete data set
  };
  
  // Save each file
  for (const [filename, data] of Object.entries(files)) {
    const filePath = path.join(config.outputDir, filename);
    await fs.promises.writeFile(filePath, JSON.stringify(data, null, 2));
    console.log(`Saved ${filePath}`);
  }
}

// Process log file
async function processLogFile() {
  console.log(`Processing log file: ${config.logFile}`);
  
  // Check if the file exists
  try {
    await fs.promises.access(config.logFile, fs.constants.F_OK);
  } catch (error) {
    console.error(`Log file does not exist: ${config.logFile}`);
    return;
  }
  
  const logs = [];
  let lineCount = 0;
  
  // Create read stream
  const fileStream = fs.createReadStream(config.logFile);
  
  // Create readline interface
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  });
  
  // Process each line
  for await (const line of rl) {
    // Skip empty lines
    if (!line.trim()) {
      continue;
    }
    
    // Parse NGINX log line
    const parsedLine = parseNginxLog(line);
    if (parsedLine) {
      logs.push(parsedLine);
      lineCount++;
      
      // Show progress
      if (lineCount % 10000 === 0) {
        console.log(`Processed ${lineCount} lines...`);
      }
      
      // Stop if we reach the sample size
      if (config.sampleSize > 0 && lineCount >= config.sampleSize) {
        break;
      }
    } else {
      console.warn(`Invalid log format: ${line}`);
    }
  }
  
  console.log(`Processing complete. Processed ${lineCount} lines.`);
  
  // Calculate and save statistics
  const stats = calculateStatistics(logs);
  await saveStatistics(stats);
  
  // Also save parsed logs to JSON file for inspection
  const parsedLogsPath = path.join(__dirname, 'parsed_logs.json');
  await fs.promises.writeFile(parsedLogsPath, JSON.stringify(logs.slice(0, 1000), null, 2));
  console.log(`Saved parsed logs to ${parsedLogsPath}`);
  
  return stats;
}

// Main function
async function main() {
  try {
    await processLogFile();
    console.log('Data generation complete!');
  } catch (error) {
    console.error('Error generating dashboard data:', error);
    process.exit(1);
  }
}

// Run the main function
main().catch(console.error);